### ISSUE TITLE
Regression in Kotlin Compatibility of SLF4JLogDelegate from 4.3.4

### ISSUE THREAD

**Issue by hjed** [eclipse-vertx/vert.x#4582]
### Version

4.3.4 onwards (tested up to 4.3.7)

### Context

When using a kotlin implementation of org.slf4j.Logger from version 4.3.7 startup fails since vert.x passes null values to varargs which isn't allowed in Kotlin. 

I think the change that caused startup to fail is:
https://github.com/eclipse-vertx/vert.x/commit/6f71b5f2adfb440b554ec9979ff3fdfd7830e985

But the solution is that https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/logging/SLF4JLogDelegate.java#L163 ideally shouldn't pass an explicitly null value to the slf4j logger (an empty array would work). 

I realise that this is really an incompatibility in the SLF4J interface, but fixing this in vert.x would be great for anyone using a Koltin implementation of the SLF4J Logger API.

### Do you have a reproducer?
A noop logger written in kotlin like the one below would fail:

```kotlin 
package org.slf4j.impl

import org.slf4j.Logger
import org.slf4j.helpers.MarkerIgnoringBase
import org.slf4j.helpers.MessageFormatter
import java.io.PrintStream
import java.util.stream.Collectors

class NoopLogger(): Logger, MarkerIgnoringBase() {

    /** Are `trace` messages currently enabled?  */
    override fun isTraceEnabled(): Boolean {
        
    }

    /**
     * A simple implementation which logs messages of level TRACE according to
     * the format outlined above.
     */
    override fun trace(msg: String) {
        
    }

    /**
     * Perform single parameter substitution before logging the message of level
     * TRACE according to the format outlined above.
     */
    override fun trace(format: String, param1: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * TRACE according to the format outlined above.
     */
    override fun trace(format: String, param1: Any?, param2: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * TRACE according to the format outlined above.
     */
    override fun trace(format: String, vararg argArray: Any?) {
        
    }

    /** Log a message of level TRACE, including an exception.  */
    override fun trace(msg: String, t: Throwable?) {
        
    }

    /** Are `debug` messages currently enabled?  */
    override fun isDebugEnabled(): Boolean {
        return true
    }

    /**
     * A simple implementation which logs messages of level DEBUG according to
     * the format outlined above.
     */
    override fun debug(msg: String) {
        
    }

    /**
     * Perform single parameter substitution before logging the message of level
     * DEBUG according to the format outlined above.
     */
    override fun debug(format: String, param1: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * DEBUG according to the format outlined above.
     */
    override fun debug(format: String, param1: Any?, param2: Any?) {
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * DEBUG according to the format outlined above.
     */
    override fun debug(format: String, vararg argArray: Any?) {
        
    }

    /** Log a message of level DEBUG, including an exception.  */
    override fun debug(msg: String, t: Throwable?) {
        
    }

    /** Are `info` messages currently enabled?  */
    override fun isInfoEnabled(): Boolean {
        return true
    }

    /**
     * A simple implementation which logs messages of level INFO according to
     * the format outlined above.
     */
    override fun info(msg: String) {
        
    }

    /**
     * Perform single parameter substitution before logging the message of level
     * INFO according to the format outlined above.
     */
    override fun info(format: String, arg: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * INFO according to the format outlined above.
     */
    override fun info(format: String, arg1: Any?, arg2: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * INFO according to the format outlined above.
     */
    override fun info(format: String, vararg argArray: Any?) {
        
    }

    /** Log a message of level INFO, including an exception.  */
    override fun info(msg: String, t: Throwable?) {
        
    }

    /** Are `warn` messages currently enabled?  */
    override fun isWarnEnabled(): Boolean {
        return true
    }

    /**
     * A simple implementation which always logs messages of level WARN
     * according to the format outlined above.
     */
    override fun warn(msg: String) {
        
    }

    /**
     * Perform single parameter substitution before logging the message of level
     * WARN according to the format outlined above.
     */
    override fun warn(format: String, arg: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * WARN according to the format outlined above.
     */
    override fun warn(format: String, arg1: Any?, arg2: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * WARN according to the format outlined above.
     */
    override fun warn(format: String, vararg argArray: Any?) {
        
    }

    /** Log a message of level WARN, including an exception.  */
    override fun warn(msg: String, t: Throwable?) {
        
    }

    /** Are `error` messages currently enabled?  */
    override fun isErrorEnabled(): Boolean =
        true
    

    /**
     * A simple implementation which always logs messages of level ERROR
     * according to the format outlined above.
     */
    override fun error(msg: String) {
        
    }

    /**
     * Perform single parameter substitution before logging the message of level
     * ERROR according to the format outlined above.
     */
    override fun error(format: String, arg: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * ERROR according to the format outlined above.
     */
    override fun error(format: String, arg1: Any?, arg2: Any?) {
        
    }

    /**
     * Perform double parameter substitution before logging the message of level
     * ERROR according to the format outlined above.
     */
    override fun error(format: String, vararg argArray: Any?) {
        f
    }

    /** Log a message of level ERROR, including an exception.  */
    override fun error(msg: String, t: Throwable?) {
        
    }

}
```

### Steps to reproduce

1. Use a logger written in kotlin
2. Start Vert.x

### Extra

* Tested this on kotlin versions 1.7.0 and 1.8.0

**Comment by hjed** [comment]
Forgot to include the relevant stack trace:

```
Caused by: java.lang.NullPointerException: Parameter specified as non-null is null: method org.slf4j.impl.GCPLogger.debug, parameter argArray
	at org.slf4j.impl.GCPLogger.debug(GCPLogger.kt)
	at io.vertx.core.logging.SLF4JLogDelegate.log(SLF4JLogDelegate.java:196)
	at io.vertx.core.logging.SLF4JLogDelegate.log(SLF4JLogDelegate.java:160)
	at io.vertx.core.logging.SLF4JLogDelegate.log(SLF4JLogDelegate.java:156)
	at io.vertx.core.logging.SLF4JLogDelegate.debug(SLF4JLogDelegate.java:122)
	at io.vertx.core.logging.LoggerFactory.<clinit>(LoggerFactory.java:37)
	... 155 more
```

**Comment by tsegismont** [comment]
Thank you for the detailed report, I will look into this asap

**Comment by tsegismont** [comment]
@hjed can you try this patch? Otherwise, can you provide a small reproducer so I can verify it?

```patch
Index: src/main/java/io/vertx/core/logging/SLF4JLogDelegate.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/io/vertx/core/logging/SLF4JLogDelegate.java b/src/main/java/io/vertx/core/logging/SLF4JLogDelegate.java
--- a/src/main/java/io/vertx/core/logging/SLF4JLogDelegate.java	(revision 85a5e412218cf102075c6331d734d6a1cdc6e22f)
+++ b/src/main/java/io/vertx/core/logging/SLF4JLogDelegate.java	(date 1673363927477)
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ * Copyright (c) 2011-2023 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License 2.0 which is available at
@@ -26,6 +26,8 @@
  */
 public class SLF4JLogDelegate implements LogDelegate {
 
+  private static final Object[] EMPTY_PARAMETERS = {};
+
   @SuppressWarnings("deprecation")
   private static final String FQCN = io.vertx.core.logging.Logger.class.getCanonicalName();
 
@@ -167,13 +169,21 @@
     // If we have both parameters and an error, we need to build a new array [params, t]
     // If we don't have parameters, we need to build a new array [t]
     // If we don't have error, it's just params.
-    Object[] parameters = params;
-    if (params != null  && t != null) {
-      parameters = new Object[params.length + 1];
-      System.arraycopy(params, 0, parameters, 0, params.length);
-      parameters[params.length] = t;
-    } else if (params == null  && t != null) {
-      parameters = new Object[] {t};
+    Object[] parameters;
+    if (params == null) {
+      if (t == null) {
+        parameters = EMPTY_PARAMETERS;
+      } else {
+        parameters = new Object[]{t};
+      }
+    } else {
+      if (t == null) {
+        parameters = params;
+      } else {
+        parameters = new Object[params.length + 1];
+        System.arraycopy(params, 0, parameters, 0, params.length);
+        parameters[params.length] = t;
+      }
     }
 
     if (logger instanceof LocationAwareLogger) {
```

**Comment by hjed** [comment]
Thank you so much for the quick reply @tsegismont.

That does indeed fix it!

**Comment by tsegismont** [comment]
Thank you for the feedback, I'll add the fix to 4.3.8

**Comment by hjed** [comment]
Brilliant, thanks!

**Comment by tsegismont** [comment]
Closed in 0137c3411
